flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/operators/join/stream/StreamingMiniBatchJoinOperatorTest.java

```diff
@@ -0,0 +1,1807 @@
+    @Test
+    public void testRightJoinWithRightArriveFirstWithNoMiniBatch() throws Exception {
+        // joinKey is LineOrd
+        // left fold  || right fold
+        testHarness.processElement2(insertRecord("Ord#X", "LineOrd#2", "AIR")); // right
+
+        testHarness.processElement1(
+                insertRecord(
+                        "Ord#1", "LineOrd#2", "3 Bellevue Drive, Pottstown, PA 19464")); // left +I
+        testHarness.processElement1(
+                insertRecord(
+                        "Ord#i",
+                        "LineOrd#6",
+                        "i6 Bellevue Drive, Pottstown, Pi 19464")); // left  +I
+
+        assertor.shouldEmit(
+                testHarness,
+                rowOfKind(RowKind.INSERT, null, null, null, "Ord#X", "LineOrd#2", "AIR"),
+                rowOfKind(RowKind.DELETE, null, null, null, "Ord#X", "LineOrd#2", "AIR"),
+                rowOfKind(
+                        RowKind.INSERT,
+                        "Ord#1",
+                        "LineOrd#2",
+                        "3 Bellevue Drive, Pottstown, PA 19464",
+                        "Ord#X",
+                        "LineOrd#2",
+                        "AIR"));
+    }
+
+    @Tag("miniBatchSize=3")
+    @Test
+    public void testFullJoinWithRightArriveFirst() throws Exception {
+        // joinKey is LineOrd
+        // left fold  || right fold
+        testHarness.processElement2(insertRecord("Ord#X", "LineOrd#2", "AIR")); // right
+
+        testHarness.processElement1(
+                insertRecord(
+                        "Ord#1", "LineOrd#2", "3 Bellevue Drive, Pottstown, PA 19464")); // left +I
+        assertor.shouldEmitNothing(testHarness);
+        testHarness.processElement1(
+                insertRecord(
+                        "Ord#i",
+                        "LineOrd#6",
+                        "i6 Bellevue Drive, Pottstown, Pi 19464")); // left  +I
+
+        assertor.shouldEmit(
+                testHarness,
+                rowOfKind(RowKind.INSERT, null, null, null, "Ord#X", "LineOrd#2", "AIR"),
+                rowOfKind(
+                        RowKind.INSERT,
+                        "Ord#i",
+                        "LineOrd#6",
+                        "i6 Bellevue Drive, Pottstown, Pi 19464",
+                        null,
+                        null,
+                        null),
+                rowOfKind(RowKind.DELETE, null, null, null, "Ord#X", "LineOrd#2", "AIR"),
+                rowOfKind(
+                        RowKind.INSERT,
+                        "Ord#1",
+                        "LineOrd#2",
+                        "3 Bellevue Drive, Pottstown, PA 19464",
+                        "Ord#X",
+                        "LineOrd#2",
+                        "AIR"));
+    }
+
+    @Tag("miniBatchSize=1")
+    @Test
+    public void testFullJoinWithRightArriveFirstWithNoMiniBatch() throws Exception {
+        // joinKey is LineOrd
+        // left fold  || right fold
```

Comment:
I think the above six tests doesn't test the fold logic, only for reducing retract message of timing-related left and right stream.

Reply:
Yes, these tests are not supposed to test the fold logic. They are only designed to test the timing-related left and right streams which could avoid emiting redundant records.